src/OmniSharp.MSBuild/ProjectManager.cs

```diff
@@ -75,6 +89,61 @@ public ProjectManager(ILoggerFactory loggerFactory, IEventEmitter eventEmitter,
             _processLoopTask = Task.Run(() => ProcessLoopAsync(_processLoopCancellation.Token));
 
             _onDirectoryFileChanged = OnDirectoryFileChanged;
+
+            if (_options.OnDemandProjectsLoad)
+            {
+                _workspace.AddWaitForProjectModelReadyEventHandler(WaitForProjectModelReadyAsync);
+            }
+        }
+
+        private async Task WaitForProjectModelReadyAsync(string documentPath)
+        {
+            // Search and queue for loading C# projects that are likely to reference the requested file.
+            // C# source files are located pretty much always in the same folder with project file that is referencing them or in a subfolder.
+            // Do not limit search by OmniSharpEnvironment.TargetDirectory to allow for loading on demand projects that are outside of it,
+            // i.e. in VSCode case potentially outside of the folder opened in the IDE.
+            string projectDir = Path.GetDirectoryName(documentPath);
+            while(projectDir != null)
+            {
+                List<string> csProjFiles = Directory.EnumerateFiles(projectDir, "*.csproj", SearchOption.TopDirectoryOnly).ToList();
+                if (csProjFiles.Count > 0)
+                {
+                    foreach(string csProjFile in csProjFiles)
+                    {
+                        if (_projectsRequestedOnDemand.TryAdd(csProjFile, 0 /*unused*/))
+                        {
+                            QueueProjectUpdate(csProjFile, allowAutoRestore:true);
+                        }
+                    }
+
+                    break;
+                }
+
+                bool foundStopAtItem = false;
+                foreach (string stopAtItemName in _options.OnDemandProjectsLoadSearchStopsAt)
```

Comment:
There's also IOmniSharpEnvironment.TargetDirectory, which is the directory that VS Code tells omnisharp it's opened against. Do you think stopping searches at TargetDirectory would sufficient? If so we could just get rid of this option.

Reply:
@rchande : I tried and found out that IOmniSharpEnvironment.TargetDirectory is quite inconsistent. It looks like the code in VSCode looks for any *.sln files and if found then uses that dir as TargetDirectory. In few bigger codebases I worked in, there were various *.sln files checked in, but I expect if I type "code ." from the root of the repo, the project search would stop at that root, not what TargetDirectory might have set to. So I think unfortunately the answer is no, the property can't be used here.

Reply:
Well...maybe. It would be perfectly reasonable to use TargetDirectory for a repo with only one sln file. Please remember that your repo with many solutions/thousands of projects/etc is an extreme outlier :)

Reply:
@rchande : this is definitely would be much simpler if OnDemandProjectsLoadSearchStopsAt is not there at all, but unfortunately I don't believe there is a way to find out where "code ." was launched from if TargetDirectory is pointing to somewhere deep in source tree. If the former path can be discovered somehow regardless to what TargetDirectory is set to then this would definitely be a good thing to switch to. Please advise.

Reply:
@dmgonch Depends on what we think the feature is. If it's "open a solution the way you normally do, but projects load more optimally", then TargetDirectory is the right solution. I think you're proposing something different though?

Reply:
@rchande : sorry for the confusion. I was referring to opening a folder, the repo root, in VSCode, which the equivalent of  executing command "code ." in the repo root folder from a command line. I  couldn't find a way to reliably get that folder path when there are *.sln files in the repo. In my testing IOmniSharpEnvironment.TargetDirectory sometimes would be set to the folder path but more often would be set of a solution file path found in the tree.

Reply:
@dmgonch Ah. In the C# extension, users can choose the "launch target" by clicking the folder icon immediately right of the green flame in the VS Code status bar at the bottom of the window. There's also a C# extension option you can set to automatically select that launch target if it's available. Do you think instructing users to use that feature would be sufficient?

Reply:
@rchande : Don't see the folder icon - what am I doing wrong? 
![image](https://user-images.githubusercontent.com/30739723/49849706-b1352180-fd8f-11e8-9b14-e61aacbffaa1.png)
Are you referring to "omnisharp.defaultLaunchSolution" setting? That one doesn't seem to match the experience for big repos. The opposite of it would be useful i.e. where any checked in solution is ignored.
Additionally there should be a way IMO to open VSCode from VS Dev Command Prompt, particularly for Net Full projects.
So to tell the truth OnDemandProjectsLoadSearchStopsAt still seems the like the best option with '.git' as a default, though in many cases it doesn't needed to be explicitly defined.
I'm truly looking to find the best compromise here but I wouldn't want to sacrifice experience working with big repos.
Please advise if you have any other ideas.

Reply:
@dmgonch That's a good idea (adding on demand project mode to the extension). That + removing the "searchatthing" will make for a more cohesive experience overall. Let's work on that once this is merged.

Reply:
To enable few more scenarios (listed in a comment in the code) I ended simply removing OnDemandProjectsLoadSearchStopsAt and going to the original implementation I had. I'm still going to be adding omnisharp.enableMsBuildLoadProjectsOnDemand C# extension setting.